### PR TITLE
feat(#9957): dynamic sorting of tasks by priority and due date

### PR DESCRIPTION
**Pull request by sookwalinga**
Closes https://github.com/medic/cht-core/issues/9957

**Is your feature request related to a problem? Please describe.**
  Currently tasks are sorted only by due date. The update takes into account both a priority score and due date.

**Describe the solution you'd like**
  Implemented a refined sorting algorithm that (Sort by priority and then due date within the same priority):
  - Primarily sorts tasks by priority (higher numbers first)
  - For tasks with equal priority, sorts by due date (earlier dates first)
  - Handles invalid values consistently:
      - Tasks with invalid priorities move to end
      - Within same priority, invalid dates move to end
      - Maintains original order when both priority and date are invalid

**Describe alternatives you've considered**
****Alternative Solution 1****
Similar to the proposed solution, but with the priority property of a task not be a number, but an array, with one number representing the priority of the task when it is not due, one number for when it’s due and one number for when it’s overdue. Then web-app will just select the right number when sorting tasks, and not recompute priority based on due date.

****Alternative Solution 2****
Options for frontend sorting: 
- Sort by due date and then priority within the same due date.
- Sorting by Due date first for tasks above the due date, then by priority for tasks that are not due. 
- Sort by priority for tasks that are above due date, then by due date for tasks that are not due.

**Additional context**
- The solution maintains backwards compatibility with existing task structures
- Includes comprehensive test cases covering edge cases like:
  - Invalid date formats (null, false, undefined).
  - Invalid priority values (strings, negative numbers).
  - Duplicate priorities and dates.
  - Missing fields.

### CODE REVIEW

**Comment by sookwalinga** [comment]
Hi @dianabarsan please find the rebased pr here

File: webapp/src/ts/reducers/tasks.ts [hunk 1]
```diff
@@ -14,20 +14,60 @@ const initialState = {
...
+    if (t?.priority === 'high') {
+      return 10;
+    } else if (t?.priority === 'medium') {
+      return 6;
+    } else if (typeof t?.priority === 'string') {
+      return 0;
+    } else {
+      return t?.priority ?? 0;
+    }
+  };
```
Comment (dianabarsan) [inline webapp/src/ts/reducers/tasks.ts:43]: I think we should not hardcode **any values**. 
If the priority is a string, we always consider it the lowest priority (for example, have this function return NaN, and check for NaN when comparing priorities.

File: webapp/src/ts/reducers/tasks.ts [hunk 2]
```diff
@@ -14,20 +14,60 @@ const initialState = {
...
+
+  const now = new Date();
+  const startOfToday = new Date(now.getFullYear(), now.getMonth(), now.getDate()).getTime();
```
Comment (dianabarsan) [inline webapp/src/ts/reducers/tasks.ts:46]: Please use `moment` functions instead of JS date.

File: webapp/src/ts/reducers/tasks.ts [hunk 3]
```diff
@@ -14,20 +14,60 @@ const initialState = {
...
+  
+  const getDateScore = (t) => {
+    if (!t?.dueDate) return Infinity;
+    const date = new Date(t.dueDate).getTime();
+    return Math.floor((date - startOfToday) / (1000 * 60 * 60 * 24));
```
Comment (dianabarsan) [inline webapp/src/ts/reducers/tasks.ts:51]: `moment` would have good date diff functionality that you can use here.

File: webapp/src/ts/reducers/tasks.ts [hunk 4]
```diff
@@ -14,20 +14,60 @@ const initialState = {
...
+  };
+
+  const p1 = getPriorityScore(t1);
+  const p2 = getPriorityScore(t2);
```
Comment (dianabarsan) [inline webapp/src/ts/reducers/tasks.ts:55]: Please don't use one letter variables. This can be `priority1`.

File: webapp/src/ts/reducers/tasks.ts [hunk 5]
```diff
@@ -14,20 +14,60 @@ const initialState = {
...
+  const days1 = getDateScore(t1);
+  const days2 = getDateScore(t2);
```
Comment (dianabarsan) [inline webapp/src/ts/reducers/tasks.ts:57]: I think `days` doesn't really convey what this value is. Maybe `dueInDays` ?

File: webapp/src/ts/services/format-date.service.ts [hunk 6]
```diff
@@ -107,14 +107,18 @@ export class FormatDateService {
...
+    if (diff < 0) {
       if (this.config.taskDaysOverdue) {
         return this.translateService.instant('task.overdue.days', { DAYS: Math.abs(diff) });
       }
 
       return this.translateService.instant('task.overdue');
     }
 
+    if (diff === 0) {
+      return this.translateService.instant('Due today');
```
Comment (dianabarsan) [inline webapp/src/ts/services/format-date.service.ts:119]: Please follow the conventions for adding new translation fields here. 
Please use a dot separated name (like `task.due.today`) and add a translation for it in all the translation files: https://github.com/medic/cht-core/tree/master/api/resources/translations

Because we don't want to change behavior for existent deployments when adding this new translation field, when adding the translation please make it the same values as the one for `task.overdue`

File: config/default/tasks.js [hunk 7]
```diff
@@ -48,8 +48,14 @@ function checkTaskResolvedForHomeVisit(contact, report, event, dueDate) {
   return isFormArraySubmittedInWindow(contact.reports, ['pregnancy_home_visit'], startTime, endTime);
 }
 
-module.exports = [
+function getPriorityCategory(score) {
```
Comment (dianabarsan) [inline config/default/tasks.js:51]: I don't think we should be adding any of these changes to the default config. Please revert them.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! I left a couple of requests below. we're getting close :)

File: webapp/tests/karma/ts/reducers/tasks.spec.ts [hunk 8]
```diff
@@ -221,7 +221,7 @@ describe('Tasks reducer', () => {
     it('should sort provided tasks by due date', () => {
       const tasks = [
         { _id: 'task1', dueDate: false, state: 'Ready', field: 1 },
-        { _id: 'task2', dueDate: undefined, state: 'Ready', field: 2 },
+        { _id: 'task2', dueDate: null, state: 'Ready', field: 2 },
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/reducers/tasks.spec.ts:224]: Is this change required? What happens if the dueDate is undefined? Does the behavior change?

File: webapp/tests/karma/ts/reducers/tasks.spec.ts [hunk 9]
```diff
@@ -242,15 +242,215 @@ describe('Tasks reducer', () => {
...
+    });
+
+    it('should sort provided tasks by priority and due date', () => {
+      const tasks = [
+        {
+          _id: 'task1',
+          dueDate: '2025-05-30',
+          priority: 'invalid',
+          state: 'Ready',
+          field: 1,
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/reducers/tasks.spec.ts:264]: Can you please add a task that has both invalid due date _and_ invalid priority? Also please don't add it as last in the initial list, it would be good to see it actually pushed down to the end.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Very nice and timely! 
I left some minor requests inline.

File: webapp/tests/karma/ts/reducers/tasks.spec.ts (hunk 8 above)
Comment (sookwalinga) [reply webapp/tests/karma/ts/reducers/tasks.spec.ts:224]: For some reason, test output shows `"dueDate": [undefined]`, yet the input data has `dueDate: undefined`. Now in the conditional statement `moment([undefined]).isValid()` returns `true ` so the test fails. I tried figuring out why `undefined` is returned as `[undefined]` but I haven't yet figured that out.

Comment (dianabarsan) [reply webapp/tests/karma/ts/reducers/tasks.spec.ts:224]: Ok. 
I believe that `moment(undefined).isValid();` is actually correct to be true, because that falls back to `moment().isValid()` which is true because moment instantiates using the current date when no input is provided. 
This means that to protect for `undefined`, you need to add a specific check in your date function.

File: config/default/tasks.js [hunk 10]
```diff
@@ -308,35 +426,58 @@ module.exports = [
...
       const endTime = addDays(dueDate, event.end + 1).getTime();
-      return isFormArraySubmittedInWindow(contact.reports, ['pnc_danger_sign_follow_up_baby'], startTime, endTime);
+      return isFormArraySubmittedInWindow(
+        contact.reports,
+        ['pnc_danger_sign_follow_up_baby'],
+        startTime,
+        endTime
+      );
+    },
+    priority: function (contact, report, event, dueDate) {
```
Comment (dianabarsan) [inline config/default/tasks.js:452]: I believe these changes should be reverted, I don't think we should be updating the default config with random priority scores. 
I will do this as part of a secondary PR once this is merged if we choose to merge this quickly.
